extension/src/documents.ts

```diff
@@ -20,16 +48,63 @@ export class DocumentManager{
 		return this.getProjectDocuments(); 
 	}
 	get ActiveDocument(): ReadonlyDocument { 
-		return vscode.window.activeTextEditor ? ReadonlyDocument.getMemoryDocument(vscode.window.activeTextEditor.document) : undefined; 
+		if (vscode.window.activeTextEditor) {
+			const key = vscode.window.activeTextEditor.document.fileName.replace(/\//g, '\\');
+			if (this._cached.has(key)){
+				const sources = this._cached.get(key);
+				if (!sources.native) {
+					sources.native = vscode.window.activeTextEditor.document;
+					sources.flags.add(Origins.OPENED);
+				}
+				if (!sources.internal || !sources.internal.equal(sources.native)){
+					sources.internal = ReadonlyDocument.getMemoryDocument(sources.native);
+				}
+				return sources.internal;
+			} else {
+				const sources = DocumentSources.create(Origins.OPENED, vscode.window.activeTextEditor.document);
+				this._cached.set(key, sources);
+				return sources.internal;
+			}
+		} else {
+			return null;
+		}		
 	}
 	
 	get ActiveTextDocument(): vscode.TextDocument { 
 		return vscode.window.activeTextEditor ? vscode.window.activeTextEditor.document :  null; 
 	}
 	get OpenedTextDocuments(): vscode.TextDocument[] { 
-		return [...this._opened.values()]; 
+		return [...this._cached.values()]
+			   .filter(p => p.native && p.flags.has(Origins.OPENED))
+			   .map(p => p.native); 
+	}
+
+	private get cacheKeys(): string [] {
+		return [...this._cached.keys()];
+	}
+	private get projectKeys(): string[] {
+		const result: string[] = [];
+		if (ProjectTreeProvider.hasProjectOpened()){
+			ProjectTreeProvider.instance().projectNode.sourceFiles.forEach(x => {
+				result.push(x.filePath.replace(/\//g, '\\'));
+			});
+		}
+		return result;
+	}
+
+	getDocument(nDoc: vscode.TextDocument) {
+		const key = nDoc.fileName.replace(/\//g, '\\');
+		if (this._cached.has(key)) {
+			const sources = this._cached.get(key);
+			sources.native = nDoc;
+			if (!sources.internal || !sources.internal.equal(sources.native)){
```

Comment:
I think this is the change we check if the atoms forest should regenerate

Reply:
Yes the `!sources.internal.equal(sources.native)` is the **cheap** check to see if there are disparities between our (internal) versions and the vscode (native) version of the document.